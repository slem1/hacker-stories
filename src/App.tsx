
import { ChangeEvent, MouseEventHandler } from 'react';
import './App.css'
import React from 'react';
import axios from 'axios';


interface Story {
  objectID: number;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
}

const App = () => {

  const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

  const useStorageState = (key: string, initialState: string): [any, any] => {
    const [value, setValue] = React.useState(localStorage.getItem(key) ?? initialState);

    React.useEffect(() => {
      localStorage.setItem(key, value);
    }, [key, value])

    return [value, setValue];
  };

  const [searchTerm, setSearchTerm] = useStorageState('search', 'React');

  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);


  type StoriesState = {
    data: Story[],
    isLoading: boolean,
    isError: boolean
  }

  type StoriesFetchInitAction = {
    type: 'STORIES_FETCH_INIT';
  };

  type StoriesFetchSuccessAction = {
    type: 'STORIES_FETCH_SUCCESS';
    payload: Story[];
  };

  type StoriesFetchFailureAction = {
    type: 'STORIES_FETCH_FAILURE';
  };

  type StoriesRemoveAction = {
    type: 'REMOVE_STORY';
    payload: Story;
  };


  type StoriesAction = StoriesFetchSuccessAction | StoriesRemoveAction | StoriesFetchInitAction | StoriesFetchFailureAction;

  const storiesReducer = (state: StoriesState, action: StoriesAction) => {
    switch (action.type) {
      case 'STORIES_FETCH_INIT':
        return { ...state, isLoading: true, isError: false };
      case 'STORIES_FETCH_SUCCESS':
        return { ...state, data: action.payload, isLoading: false, isError: false };
      case 'STORIES_FETCH_FAILURE':
        return { ...state, isLoading: false, isError: true }
      case 'REMOVE_STORY':
        return {
          ...state, data: state.data.filter(story => action.payload.objectID !== story.objectID),
        };
      default:
        throw new Error();
    }
  }

  const [stories, dispatchStories] = React.useReducer(storiesReducer, { data: [], isLoading: false, isError: false });

  const handleFetchStories = React.useCallback(async () => {

    dispatchStories({ type: 'STORIES_FETCH_INIT' });


    try {
      const result = await axios.get(url);      
          dispatchStories({
            type: 'STORIES_FETCH_SUCCESS',
            payload: result.data.hits
          });   
    } catch {
      dispatchStories({type: 'STORIES_FETCH_FAILURE'});
    }
  
  }, [url])

  React.useEffect(() => { handleFetchStories();}, [handleFetchStories]);

  const handleSearchInput = (event : ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  }
  
  const handleSearchSubmit = () => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
  }

  const handleRemoveStory = (item: Story) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item
    });
  }

  return (<div>

    <h1>My Hacker Stories</h1>

    <InputWithLabel id="search" value={searchTerm} onInputChange={handleSearchInput} isFocused>
      Search:
    </InputWithLabel>

    <button type="button" disabled={!searchTerm} value={searchTerm} onClick={handleSearchSubmit}>Search</button>

    <hr />
    {stories.isError && <p>Something went wrong ...</p>}

    {stories.isLoading ? <p>Loading ... </p> : <List list={stories.data} onRemoveItem={handleRemoveStory} />}

  </div>
  )
}

interface ListProps {
  list: Story[],
  onRemoveItem: (item: Story) => void
}

interface ItemProps {
  item: Story,
  onRemoveItem: (item: Story) => void
}

interface SearchProps {
  onSearch: (event: ChangeEvent<HTMLInputElement>) => void
  search: string
}

interface InputWithLabelProps {
  id: string,
  value: string,
  type?: string,
  children: React.ReactNode,
  isFocused: boolean,
  onInputChange: (event: ChangeEvent<HTMLInputElement>) => void
}

const InputWithLabel = ({ id, value, type = 'text', children, isFocused, onInputChange }: InputWithLabelProps) => {

  //A
  const inputRef = React.useRef<HTMLInputElement>(null);

  //C
  React.useEffect(() => {
    //only if isFocused and ref.current has been set
    if (isFocused && inputRef.current) {
      //D
      inputRef.current.focus();
    }
  }, [isFocused])

  return (
    <>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      {/* B */}
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange} />
    </>
  )
}


const List = ({ list, onRemoveItem }: ListProps) => {

  return (<div>
    <ul>
      {list.map((item) => (
        <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
      ))}
    </ul>
  </div>
  )
}



const Item = ({ item, onRemoveItem }: ItemProps
) => {
  return (
    <li>
      <span>
        <a href={item.url}>{item.title}</a>
      </span>
      <span>{item.author}</span>
      <span>{item.num_comments}</span>
      <span>{item.points}</span>
      <button key={item.objectID} value={item.objectID} onClick={() => onRemoveItem(item)}>Dismiss</button>
    </li>
  );
}

export default App
