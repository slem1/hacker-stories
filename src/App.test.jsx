import {
    fireEvent,
    render,
    screen,
    waitFor,
} from '@testing-library/react';
import axios from 'axios';
import { describe, expect, it, vi } from 'vitest';
import App, { Item, SearchForm, storiesReducer } from './App';

vi.mock('axios');


const storyOne = {
    title: 'React',
    url: 'https://react.dev/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
};

const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};

const stories = [storyOne, storyTwo];


describe('storiesReducer', () => {
    it('remove a story from all stories', () => {

        const state = { data: stories, isLoading: false, isError: false };
        const action = { type: 'REMOVE_STORY', payload: storyOne }

        const newState = storiesReducer(state, action);

        const expectedState = { data: [storyTwo], isLoading: false, isError: false };

        expect(newState).toStrictEqual(expectedState);

    });
});

describe('Item', () => {
    it('renders all properties', () => {

        render(<Item item={storyOne}></Item>);

        screen.debug();

        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute('href', 'https://react.dev/');
    });

    it('clicking the dismiss button calls the callback handler', () => {
        const handleRemoveItem = vi.fn()

        render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

        fireEvent.click(screen.getByRole('button'));

        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    })
});

describe('SearchForm', () => {

    const searchFormProps = {
        searchTerm: 'React',
        onSearchInput: vi.fn(),
        searchAction: vi.fn()
    };

    it('renders the input field with its value', () => {
        render(<SearchForm {...searchFormProps} />)

        screen.debug();

        expect(screen.getByDisplayValue('React')).toBeInTheDocument();

    });

    it('renders the correct label', () => {
        render(<SearchForm {...searchFormProps} />)

        expect(screen.getByLabelText(/Search/)).toBeInTheDocument();

    });

    it('calls onSearchInput on input field change', () => {

        render(<SearchForm {...searchFormProps} />);

        fireEvent.change(screen.getByDisplayValue('React'), { target: { value: 'Redux' } });

        expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
    });

    it('calls searchAction button submit click', () => {

        render(<SearchForm {...searchFormProps} />);

        fireEvent.click(screen.getByRole('button'))

        expect(searchFormProps.searchAction).toHaveBeenCalledTimes(1);
    });
})

describe('App', () => {
    it('succeeds fetching data', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories
            }
        });

        axios.get.mockImplementationOnce(() => promise);

        render(<App />);

        screen.debug();

        expect(screen.queryByText(/Loading/)).toBeInTheDocument();

        await waitFor(async () => await promise);
        screen.debug();

        expect(screen.queryByText(/Loading/)).toBeNull();

        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
        expect(screen.getAllByText('Dismiss').length).toBe(2);
    })

    it('fails fetching data', async () => {
        const promise = Promise.reject();

        axios.get.mockImplementationOnce(() => promise);

        render(<App />)

        expect(screen.getByText(/Loading/)).toBeInTheDocument();

        try {
            await waitFor(async () => await promise);
        } catch (error) {
            expect(screen.queryByText(/Loading/)).toBeNull();
            expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
        }
    })

    it('remove a story', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });
        axios.get.mockImplementationOnce(() => promise);
        render(<App />);
        await waitFor(async () => await promise);
        expect(screen.getAllByText('Dismiss').length).toBe(2);
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

        fireEvent.click(screen.getAllByText('Dismiss')[0]);

        expect(screen.getAllByText('Dismiss').length).toBe(1);
        expect(screen.queryByText('Jordan Walke')).toBe(null);

    });

})

describe('snapshot', () => {

    const searchFormProps = {
        searchTerm: 'React',
        onSearchInput: vi.fn(),
        searchAction: vi.fn()
    };

    it('renders a snapshot', () => {
        const { container } = render(<SearchForm {...searchFormProps} />);
        expect(container.firstChild).toMatchSnapshot();

    })

})