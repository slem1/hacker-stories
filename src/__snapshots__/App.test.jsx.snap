// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot > renders a snapshot 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="search-form"
>
  <label
    class="label"
    for="search"
  >
    Search:
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button button_large"
    type="submit"
  >
    Submit
  </button>
</form>
`;
